src/reducers/playground.ts

```diff
@@ -1,28 +1,106 @@
-import { Action, Reducer } from 'redux'
-import { IUpdateEditorValue, UPDATE_EDITOR_VALUE } from '../actions/playground'
+import { Reducer } from 'redux'
+import {
+  CLEAR_CONTEXT,
+  CLEAR_REPL_INPUT,
+  CLEAR_REPL_OUTPUT,
+  EVAL_INTERPRETER_ERROR,
+  EVAL_INTERPRETER_SUCCESS,
+  HANDLE_CONSOLE_LOG,
+  IAction,
+  SEND_REPL_INPUT_TO_OUTPUT,
+  UPDATE_EDITOR_VALUE,
+  UPDATE_REPL_VALUE
+} from '../actions/actionTypes'
+import { createContext } from '../slang'
+import { CodeOutput, defaultPlayground, InterpreterOutput, IPlaygroundState } from './states'
 
-export interface IPlaygroundState {
-  editorValue: string
-}
-
-/**
- * The default (initial) state of the `IPlaygroundState`
- */
-const defaultState: IPlaygroundState = {
-  editorValue: ''
-}
-
-/**
- * The reducer for `IPlaygroundState`
- *
- * UPDATE_EDITOR_VALUE: Update the `editorValue` property
- */
-export const reducer: Reducer<IPlaygroundState> = (state = defaultState, action: Action) => {
+export const reducer: Reducer<IPlaygroundState> = (state = defaultPlayground, action: IAction) => {
+  let newOutput: InterpreterOutput[]
+  let lastOutput: InterpreterOutput
   switch (action.type) {
     case UPDATE_EDITOR_VALUE:
       return {
         ...state,
-        editorValue: (action as IUpdateEditorValue).payload
+        editorValue: action.payload
+      }
+    case UPDATE_REPL_VALUE:
+      return {
+        ...state,
+        replValue: action.payload
+      }
+    case CLEAR_REPL_INPUT:
+      return {
+        ...state,
+        replValue: ''
+      }
+    case CLEAR_REPL_OUTPUT:
+      return {
+        ...state,
+        output: []
+      }
+    case CLEAR_CONTEXT:
+      return {
+        ...state,
+        context: createContext()
+      }
+    case HANDLE_CONSOLE_LOG:
```

Comment:
Not sure what's the benefit of grouping console logs. In fact, under certain circumstances this might not be what you want (e.g what if the console log is done asynchronously?)

Reply:
The intent would be to group the `display()` outputs of a particular program together in one Card, as compared to one card per output (as in sourceacademy), so as to increase the readability of one's program outputs (code output, displays and then the result). 

I did not consider asynchronous outputs previously, it could cause problems with the solution here. @ningyuansg any thoughts?

Reply:
We are trying to have one output per program. That is, each program results in exactly one card being displayed (with the exception of CodeOutput). 

I recorded a comparison [here](https://youtu.be/8RArdxjIuHA). The distinction between logs and results/errors will be more prominent after we add text colour with code tags.

The aim is to emulate normal behaviour in most shells. When a program is run, the user is only given back the prompt after stdout/stderr has finished printing their stuff. The area of output between prompts (`user $`) is analogous to our `<Card />` block elements.

Async certainly deviates from this behaviour, but the reducer is smart enough to handle it at least without crashing. Async is also pretty rare in CS1101S. It was present only as a 6th MC sidequest last year, and considering next year's CS1101S is 4 MCs the source language may not need to support async.